core/src/Kokkos_CopyViews.hpp

```diff
@@ -3278,26 +3278,25 @@ impl_realloc(Kokkos::View<T, P...>& v, const size_t n0, const size_t n1,
   const bool sizeMismatch = Impl::size_mismatch(v, v.rank_dynamic, new_extents);
 
   if (sizeMismatch) {
-    v = view_type();  // Deallocate first, if the only view to allocation
-    v = view_type(arg_prop, n0, n1, n2, n3, n4, n5, n6, n7);
+    auto const& label = v.label();  // Save the label
+    v = view_type();  // Best effort to deallocate in case no other view refers
+                      // to the shared allocation
+    using alloc_prop = Impl::ViewCtorProp<ViewCtorArgs..., std::string>;
+    alloc_prop arg_prop_copy(arg_prop);
+    static_cast<Kokkos::Impl::ViewCtorProp<void, std::string>&>(arg_prop_copy)
+        .value = label;
+    v          = view_type(arg_prop_copy, n0, n1, n2, n3, n4, n5, n6, n7);
   } else if (alloc_prop_input::initialize) {
     if (alloc_prop_input::has_execution_space) {
-      // Add execution_space if not provided to avoid need for if constexpr
       using alloc_prop = Impl::ViewCtorProp<
           ViewCtorArgs...,
           std::conditional_t<alloc_prop_input::has_execution_space,
                              std::integral_constant<unsigned int, 2>,
-                             typename view_type::execution_space>,
-          std::string>;
+                             typename view_type::execution_space>>;
       alloc_prop arg_prop_copy(arg_prop);
-      static_cast<Kokkos::Impl::ViewCtorProp<void, std::string>&>(arg_prop_copy)
-          .value                 = v.label();
-      using execution_space_type = typename alloc_prop::execution_space;
-      const execution_space_type& exec_space =
-          static_cast<
-              Kokkos::Impl::ViewCtorProp<void, execution_space_type> const&>(
-              arg_prop_copy)
-              .value;
+      auto const& exec_space = static_cast<Kokkos::Impl::ViewCtorProp<
+          void, typename alloc_prop::execution_space> const&>(arg_prop_copy)
+                                   .value;
       Kokkos::deep_copy(exec_space, v, typename view_type::value_type{});
```

Comment:
#5314 will clean this up some more since we don't really need to modify `arg_prop` when we can use `if constexpr`.